### PR TITLE
Session 1

### PR DESCRIPTION
**Pull request by vteraz**

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -4,3 +4,16 @@ name := "lohika-scala-camp"
...
+resolvers ++= Seq(
+  "bintray-sbt-plugin-releases" at "http://dl.bintray.com/content/sbt/sbt-plugin-releases",
+  "Typesafe repository" at "http://repo.typesafe.com/typesafe/releases/",
+)
+
+libraryDependencies += "org.scalatest" %% "scalatest" % "3.0.5" % "test"
+libraryDependencies +=  "org.scalaj" %% "scalaj-http" % "2.4.1"
+libraryDependencies += "com.typesafe.scala-logging" %% "scala-logging" % "3.9.2"
+libraryDependencies += "ch.qos.logback" % "logback-classic" % "1.2.3"
+libraryDependencies += "org.scalamock" %% "scalamock" % "4.1.0" % Test
```
Comment (azhur) [inline build.sbt]: why you need all of that? `scalatest` is all you need for that task.

File: src/main/scala/com/lohika/vteraz/Main.scala [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    * @param action        Action to perform
+    * @param acceptResults Predicate to test whether result of action is valid
+    * @param retries       Number of tries and timeout between them
+    */
+  @tailrec
+  def safeRetry[A](action: () => A,
+                   acceptResults: A => Boolean,
+                   retries: List[FiniteDuration]): A = {
+    retries match {
+      case Nil => throw new FailedToPerformActionException() // this should never be happened
```
Comment (azhur) [inline src/main/scala/com/lohika/vteraz/Main.scala]: if this should never happen why it is there? What happens if i call this function and pass empty list there?

File: src/main/scala/com/lohika/vteraz/Main.scala [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    * Performs action specified amount of times and try to get suitable result.
+    *
+    * <b>If no suitable result found - returns last result of action no matter whether it's correct</b>
+    *
+    * @param action        Action to perform
+    * @param acceptResults Predicate to test whether result of action is valid
+    * @param retries       Number of tries and timeout between them
+    */
+  @tailrec
+  def safeRetry[A](action: () => A,
```
Comment (azhur) [inline src/main/scala/com/lohika/vteraz/Main.scala]: The task requires to implement 1 function, not many.

File: src/main/scala/com/lohika/vteraz/Main.scala [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    * @param acceptResults Predicate to test whether result of action is valid
+    * @param retries       Number of tries and timeout between them
+    * @throws FailedToPerformActionException In case no valid result achieved
+    */
+  @tailrec
+  def retry[A](action: () => A,
+               acceptResults: A => Boolean,
+               retries: List[FiniteDuration]): A = {
+    retries match {
+      case Nil => throw new FailedToPerformActionException()
```
Comment (azhur) [inline src/main/scala/com/lohika/vteraz/Main.scala]: why? what if i don't want to execute retries at all? it should work for empty lists too

File: src/main/scala/com/lohika/vteraz/Main.scala [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+  /**
+    * Performs action specified amount of times and try to get suitable result.
+    *
+    * <b>If no suitable result found - throws exception</b>
+    *
+    * @param action        Action to perform
+    * @param acceptResults Predicate to test whether result of action is valid
+    * @param retries       Number of tries and timeout between them
+    * @throws FailedToPerformActionException In case no valid result achieved
```
Comment (azhur) [inline src/main/scala/com/lohika/vteraz/Main.scala]: the retry function shouldn't throw any exceptions.

File: build.sbt (hunk 1 above)
Comment (vteraz) [reply build.sbt]: just tried how it works with `sbt`

**Comment by vteraz** [comment]
@azhur I've deleted all useless code. 
Regarding exception in `retry`: I thought it would be nice to know if all retries was failed. Anyway, I get rid of it.  

thanks for comments and hope you'll have a chance to check PR again.

**Pull request merged**